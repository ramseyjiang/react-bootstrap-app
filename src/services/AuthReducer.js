import {
  getLocal,
  removeLocal,
  setLocal,
} from "../components/common/utils/Storage.js";

export const initAuth = {
  isLoggedIn: false,
  loading: false,
  username: null,
  role: { isAdmin: false, isManager: false, isStaff: false, isClient: true }, //it is used to extension the roles management.
  error: null,
};

// Actions
export const SUCCESS = "SUCCESS";
export const FAIL = "FAIL";
export const LOGOUT = "LOGOUT";
export const LOADING = "LOADING";
export const INIT = "INIT";

export function authReducer(state, action) {
  switch (action.type) {
    case INIT:
      return initAccess();
    case LOADING:
      return { ...state, loading: true };
    case SUCCESS:
      return loginSuccess(action);
    case FAIL:
      return loginFailure(action);
    case LOGOUT:
      return logout();
    default:
      return state;
  }
}

const initAccess = () => {
  let username = getLocal("username");
  return { ...initAuth, isLoggedIn: username !== null, username: username };
};

const loginSuccess = (action) => {
  setLocal("username", action.username);
  return { ...initAuth, isLoggedIn: true, username: action.username };
};

const loginFailure = (action) => {
  setLocal("username", null);
  return { ...initAuth, error: action.error };
};

const logout = () => {
  removeLocal("username");
  return { ...initAuth };
};
