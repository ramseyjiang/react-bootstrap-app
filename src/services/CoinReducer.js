// Actions
export const LOADING = "LOADING";
export const GET_COIN_SUCCESS = "GET_COIN_SUCCESS";
export const GET_COIN_FAILURE = "GET_COIN_FAILURE";

export const initState = {
  data: {
    btc: "",
    ltc: "",
    eth: "",
  },
  loading: false,
  errorMessage: null,
};

export const coinReducer = (state, action) => {
  switch (action.type) {
    case LOADING: {
      return { ...state, loading: true };
    }
    case GET_COIN_SUCCESS: {
      return {
        ...initState,
        data: {
          btc: action.data.BTC.USD,
          ltc: action.data.LTC.USD,
          eth: action.data.ETH.USD,
        },
      };
    }
    case GET_COIN_FAILURE: {
      return {
        ...initState,
        errorMessage: "Something wrong!",
      };
    }
    default:
      throw new Error();
  }
};
