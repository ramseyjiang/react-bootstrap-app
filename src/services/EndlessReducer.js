// Actions
export const LOADING = "LOADING";
export const GET_DATA_SUCCESS = "GET_DATA_SUCCESS";
export const GET_DATA_FAILURE = "GET_DATA_FAILURE";
export const SET_LOAD_MORE = "SET_LOAD_MORE";

export const initState = {
  data: [],
  loading: false,
  errorMessage: null
};

export const endlessReducer = (state, action) => {
  switch (action.type) {
    case LOADING: {
      return { ...state, loading: true };
    }
    case GET_DATA_SUCCESS: {
      return {
        ...initState,
        data: action.data
      };
    }
    case GET_DATA_FAILURE: {
      return {
        ...initState,
        errorMessage: "Something wrong!"
      };
    }
    case SET_LOAD_MORE: {
      return {
        ...initState,
        data: [...state.data, ...action.data]
      };
    }
    default:
      throw new Error();
  }
};
