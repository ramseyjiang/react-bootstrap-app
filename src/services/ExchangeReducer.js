export const LOADING = "LOADING";
export const GET_LATEST_SUCCESS = "GET_LATEST_SUCCESS";
export const GET_LATEST_FAILURE = "GET_LATEST_FAILURE";

const CURRENCIES = ["AUD", "CNY", "NZD", "SGD"];

export const initState = {
  data: {
    initialized: false,
    rates: {},
    base: "USD",
    tableKeys: {
      AUD: "AUD",
      CNY: "CNY",
      NZD: "NZD",
      SGD: "SGD"
    },
    barData: {}
  },
  loading: false,
  errorMessage: null
};

export const exchangeReducer = (state, action) => {
  switch (action.type) {
    case LOADING: {
      return { ...state, loading: true };
    }
    case GET_LATEST_SUCCESS: {
      return {
        ...initState,
        data: {
          initialized: true,
          rates: action.data.rates,
          base: action.data.base,
          barData: generateBarData(action.data.rates),
          tableKeys: initState.data.tableKeys
        }
      };
    }
    case GET_LATEST_FAILURE: {
      return {
        ...initState,
        errorMessage: "Something wrong!"
      };
    }
    default:
      throw new Error();
  }
};

const generateBarData = rates => {
  let filteredRates = Object.keys(rates).filter(key =>
    CURRENCIES.includes(key)
  );

  return {
    labels: filteredRates,
    datasets: [
      {
        backgroundColor: "green",
        data: filteredRates.map(key => rates[key])
      }
    ]
  };
};
