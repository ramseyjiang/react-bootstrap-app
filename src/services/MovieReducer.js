export const initState = {
  data: [],
  loading: false,
  errorMessage: null
};

export const moviesReducer = (state, action) => {
  switch (action.type) {
    case "loading": {
      return { ...state, loading: true };
    }
    case "success": {
      return {
        data: action.data,
        loading: false,
        errorMessage: null
      };
    }
    case "failure": {
      return {
        data: [],
        loading: false,
        errorMessage: action.error
      };
    }
    default:
      throw new Error();
  }
};
