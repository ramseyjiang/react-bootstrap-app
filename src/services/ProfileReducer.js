// Actions
export const LOADING = "LOADING";
export const GET_PROFILE_SUCCESS = "GET_PROFILE_SUCCESS";
export const GET_PROFILE_FAILURE = "GET_PROFILE_FAILURE";

export const initState = {
  data: null,
  loading: false,
  errorMessage: null,
};

export const profileReducer = (state, action) => {
  switch (action.type) {
    case LOADING: {
      return { ...state, loading: true };
    }
    case GET_PROFILE_SUCCESS: {
      return {
        ...initState,
        data: action.data,
      };
    }
    case GET_PROFILE_FAILURE: {
      return {
        ...initState,
        errorMessage: "Something wrong!",
      };
    }
    default:
      throw new Error();
  }
};
