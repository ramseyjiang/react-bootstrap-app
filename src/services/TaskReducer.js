import uniqid from "uniqid";

export const initState = {
  counter: 2,
  tasks: [
    {
      id: 1,
      title: "Default task1",
      creator: "ramsey"
    },
    {
      id: 2,
      title: "Default task2",
      creator: "davy"
    }
  ], //originalTasks is used to show the original data after click search button.
  originalTasks: [
    {
      id: 1,
      title: "Default task1",
      creator: "ramsey"
    },
    {
      id: 2,
      title: "Default task2",
      creator: "davy"
    }
  ]
};

export const tasksReducer = (state, action) => {
  switch (action.type) {
    case "add": {
      return addTask(state, action);
    }

    case "delete": {
      return deleteTask(state, action);
    }

    case "search": {
      return searchTask(state, action);
    }

    case "reload": {
      return reloadTask(state, action);
    }
    default:
      return state;
  }
};

const addTask = (state, action) => {
  const newCounter = state.counter + 1;

  const newTask = {
    id: uniqid(),
    title: action.task.title,
    creator: action.task.creator
  };

  return {
    counter: newCounter,
    tasks: [...state.tasks, newTask],
    originalTasks: [...state.tasks, newTask]
  };
};

const deleteTask = (state, action) => {
  return {
    counter: state.counter - 1,
    tasks: state.tasks.filter(task => {
      return task.id !== action.taskId;
    }),
    originalTasks: state.originalTasks.filter(task => {
      return task.id !== action.taskId;
    })
  };
};

const searchTask = (state, action) => {
  let tasks =
    state.originalTasks.length > state.tasks.length
      ? (state.tasks = state.originalTasks)
      : state.tasks;

  tasks = action.query.length
    ? tasks.filter(task => {
        return (
          task.title.includes(action.query) ||
          task.creator.includes(action.query)
        );
      })
    : tasks;

  return {
    counter: tasks.length,
    tasks: tasks,
    originalTasks: state.originalTasks
  };
};

const reloadTask = (state, action) => {
  return {
    counter: state.originalTasks.length,
    tasks: state.originalTasks,
    originalTasks: state.originalTasks
  };
};
