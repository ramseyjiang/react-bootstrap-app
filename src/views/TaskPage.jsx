import React from "react";
import TaskContextProvider from "../contexts/TaskContext";
import TaskForm from "../components/tasks/TaskForm";
import TaskList from "../components/tasks/TaskList";
import TaskSearch from "../components/tasks/TaskSearch";

const TaskPage = () => {
  return (
    <TaskContextProvider>
      <TaskSearch />
      <TaskList />
      <TaskForm />
    </TaskContextProvider>
  );
};

export default TaskPage;
