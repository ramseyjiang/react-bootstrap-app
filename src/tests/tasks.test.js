import React from "react";
import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import TaskPage from "../views/TaskPage";
import { initState } from "../services/TaskReducer";

test("show task search component", () => {
  const { getByPlaceholderText } = render(<TaskPage />);
  expect(getByPlaceholderText(/Input search content/i)).toBeInTheDocument();
});

test("show task list component", () => {
  const { getByText } = render(<TaskPage />);
  expect(getByText(/Total tasks is/i)).toBeInTheDocument();
});

test("show task form component", () => {
  const { getByPlaceholderText } = render(<TaskPage />);
  expect(getByPlaceholderText(/Please input title/i)).toBeInTheDocument();
  expect(getByPlaceholderText(/Please input creator/i)).toBeInTheDocument();
});

test("snapshot renders correctly", () => {
  const tree = renderer.create(<TaskPage />).toJSON();
  expect(tree).toMatchSnapshot();
});

//todo test click search button, click refresh button, edit button, delete button, add task button
