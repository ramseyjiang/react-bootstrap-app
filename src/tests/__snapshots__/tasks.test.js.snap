// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot renders correctly 1`] = `
Array [
  <div
    className="container"
  >
    <form
      className=""
    >
      <div
        className="form-group row"
      >
        <div
          className="col-sm-8"
        >
          <input
            className="form-control"
            onChange={[Function]}
            placeholder="Input search content"
            type="text"
            value=""
          />
        </div>
        <div
          className="col-sm-4"
        >
          <button
            className="btn btn-primary btn-sm"
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            Search
          </button>
           
          <button
            className="btn btn-success btn-sm"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Refresh
          </button>
        </div>
      </div>
    </form>
  </div>,
  <div
    className="container"
  >
    <h6
      md={
        Object {
          "offset": 6,
        }
      }
    >
      Total tasks is 
      2
    </h6>
    <table
      className="table"
    >
      <thead>
        <tr>
          <th>
            Task Title
          </th>
          <th>
            Task Creator
          </th>
          <th>
            Operation
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Default task1
             
             
            <button
              className="btn btn-warning btn-sm"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Edit
            </button>
          </td>
          <td>
            ramsey
          </td>
          <td>
            <button
              className="btn btn-danger btn-sm"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td>
            Default task2
             
             
            <button
              className="btn btn-warning btn-sm"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Edit
            </button>
          </td>
          <td>
            davy
          </td>
          <td>
            <button
              className="btn btn-danger btn-sm"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  <div
    className="container"
  >
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="form-row"
      >
        <div
          className="form-group col"
        >
          <label
            className="form-label"
          >
            Title
          </label>
          <input
            className="form-control"
            name="title"
            onChange={[Function]}
            placeholder="Please input title"
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group col"
        >
          <label
            className="form-label"
          >
            Creator
          </label>
          <input
            className="form-control"
            name="creator"
            onChange={[Function]}
            placeholder="Please input creator"
            required={true}
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        className="btn btn-primary btn-sm"
        disabled={false}
        type="submit"
      >
        Add Task
      </button>
    </form>
  </div>,
]
`;
